### ISSUE TITLE
TypeError: no implicit conversion of Fixnum into String

### ISSUE THREAD

**Issue by dhoer** [opscode-cookbooks/homebrew#71]
Homebrew for Chef 12.3.0 on Mac OS X 10.10.3 throws `TypeError: no implicit conversion of Fixnum into String`.  But homebrew works for Chef 12.0.3.

Stacktrace:
```
     Starting Chef Client, version 12.3.0
       Compiling Cookbooks...
       Recipe: chrome::dmg
        (up to date)
         Converging 14 resources
       Recipe: homebrew::default
        (up to date)
         * execute[install homebrew] action run (skipped due to not_if)
        (skipped due to not_if)

           - execute /usr/local/bin/brew update || true
       Recipe: homebrew::cask
         * homebrew_tap[caskroom/cask] action tap (up to date)


           ================================================================================
           Error executing action `install` on resource 'homebrew_package[brew-cask]'
           ================================================================================

           TypeError
           ---------
           no implicit conversion of Fixnum into String

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `getpwnam'
           /tmp/kitchen/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `get_response_from_command'
           /tmp/kitchen/cookbooks/homebrew/libraries/homebrew_package.rb:70:in `brew'
           /tmp/kitchen/cookbooks/homebrew/libraries/homebrew_package.rb:50:in `install_package'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cookbooks/homebrew/recipes/cask.rb

            23: package 'brew-cask'
            24:

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cookbooks/homebrew/recipes/cask.rb:23:in `from_file'

           homebrew_package("brew-cask") do
             action :install
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             package_name "brew-cask"
             version "0.54.0"
             timeout 900
             declared_type :package
             cookbook_name :homebrew
             recipe_name "cask"
           end


       Running handlers:
       [2015-05-20T22:07:24-07:00] ERROR: Running exception handlers
       Running handlers complete
       [2015-05-20T22:07:24-07:00] ERROR: Exception handlers complete
       [2015-05-20T22:07:24-07:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       Chef Client failed. 1 resources updated in 12.885101 seconds
       [2015-05-20T22:07:24-07:00] ERROR: homebrew_package[brew-cask] (homebrew::cask line 23) had an error: TypeError: no implicit conversion of Fixnum into String
       [2015-05-20T22:07:24-07:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by kbruner** [comment]
Same problem here, although oddly, "homebrew_tap 'caskroom/cask'" runs fine, as do other formulae.  However, the mongo24 formula throws the error.

Using chef-client 12.3.0 from the latest Mac OS X ChefDK dmg, on OS X 10.10.1.  Works fine with chef 11.18.1.

TypeError: homebrew_package[mongodb24] (homebrew::install_formulas line 23) had an error: TypeError: no implicit conversion of Fixnum into String
/tmp/chef/local-mode-cache/cache/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `getpwnam'
/tmp/chef/local-mode-cache/cache/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `get_response_from_command'
/tmp/chef/local-mode-cache/cache/cookbooks/homebrew/libraries/homebrew_package.rb:70:in `brew'
/tmp/chef/local-mode-cache/cache/cookbooks/homebrew/libraries/homebrew_package.rb:50:in `install_package'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/package.rb:87:in `block in action_install'
/opt/chefdk/embedded/apps/chef/lib/chef/mixin/why_run.rb:52:in `call'
/opt/chefdk/embedded/apps/chef/lib/chef/mixin/why_run.rb:52:in `add_action'
/opt/chefdk/embedded/apps/chef/lib/chef/provider.rb:165:in `converge_by'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/package.rb:86:in `action_install'
/opt/chefdk/embedded/apps/chef/lib/chef/provider.rb:130:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/resource.rb:562:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:339:in `block in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `catch'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:353:in `converge_and_save'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:457:in `run'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:271:in `block in fork_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:225:in `block in run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:375:in `run_application'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:60:in `run'
/opt/chefdk/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
/opt/chefdk/bin/chef-client:51:in `load'
/opt/chefdk/bin/chef-client:51:in `<main>'

**Comment by kbruner** [comment]
This line in libraries/homebrew_mixin.rb seems to be causing the problem:

          @homebrew_owner ||= Chef12HomebrewUser.new.find_homebrew_uid

It's returning the uid, whereas everything else returns the username string, and Etc.getpwnam called in libraries/homebrew_packages.rb expects the username string.

**Comment by dhoer** [comment]
Nice find! :+1:

**Comment by trinitronx** [comment]
Good find!  I was able to reproduce the error in a `pry` session (thanks to @jtimberman's great [blog post][2], and verify that the `Chef12HomebrewUser.find_homebrew_uid` method returning a UID is the issue because it gets passed into `Etc.getpwnam` which expects a string.

Looks like the fix in PR #72 will fix it!


For Googlers...

I ran into this issue also with the [`sprout-homebrew`][1] cookbook.  It appears to depend on this official homebrew cookbook for the default `package` resource override functionality.  That is to say: it requires this cookbook so that for `<= Chef 11` it will use homebrew as the default package manager (provider for the `package` resource).  The error arises when it tries to use homebrew to install `brew-cask`:

```
include_recipe 'homebrew'

directory '/usr/local/Library/Taps' do
  owner node['sprout']['user']
  recursive true
end

execute 'tap cask' do
  command 'brew tap "caskroom/homebrew-cask"'
  user node['sprout']['user']
end

package 'brew-cask'

package 'brew-cask' do
  action :upgrade
  ignore_failure true
end
```

The errors looked like this:

```
Recipe: sprout-homebrew::cask
  * directory[/usr/local/Library/Taps] action create[2015-06-07T18:24:04-06:00] INFO: Processing directory[/usr/local/Library/Taps] action create (sprout-homebrew::cask line 3)
 (up to date)
  * execute[tap cask] action run[2015-06-07T18:24:04-06:00] INFO: Processing execute[tap cask] action run (sprout-homebrew::cask line 8)
==> Tapping caskroom/cask
Cloning into '/usr/local/Library/Taps/caskroom/homebrew-cask'...
Tapped 1 formula (2958 files, 23M)
It looks like you tapped a private repository. To avoid entering your
credentials each time you update, you can use git HTTP credential
caching or issue the following command:

  cd /usr/local/Library/Taps/caskroom/homebrew-cask
  git remote set-url origin git@github.com:caskroom/homebrew-cask.git
[2015-06-07T18:24:07-06:00] INFO: execute[tap cask] ran successfully

    - execute brew tap "caskroom/homebrew-cask"
  * homebrew_package[brew-cask] action install[2015-06-07T18:24:07-06:00] INFO: Processing homebrew_package[brew-cask] action install (sprout-homebrew::cask line 13)

    
    ================================================================================
    Error executing action `install` on resource 'homebrew_package[brew-cask]'
    ================================================================================
    
    TypeError
    ---------
    no implicit conversion of Fixnum into String
    
    Cookbook Trace:
    ---------------
    /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `getpwnam'
    /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/homebrew/libraries/homebrew_package.rb:102:in `get_response_from_command'
    /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/homebrew/libraries/homebrew_package.rb:70:in `brew'
    /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/homebrew/libraries/homebrew_package.rb:50:in `install_package'
    
    Resource Declaration:
    ---------------------
    # In /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/sprout-homebrew/recipes/cask.rb
    
     13: package 'brew-cask'
     14: 
    
    Compiled Resource:
    ------------------
    # Declared in /Users/jcuzella/src/pub/soloist/sprout-wrap/cookbooks/sprout-homebrew/recipes/cask.rb:13:in `from_file'
    
    homebrew_package("brew-cask") do
      action :install
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      package_name "brew-cask"
      version "0.54.0"
      timeout 900
      declared_type :package
      cookbook_name :"sprout-homebrew"
      recipe_name "cask"
    end
    
[2015-06-07T18:24:08-06:00] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2015-06-07T18:24:08-06:00] ERROR: Running exception handlers
Running handlers complete
[2015-06-07T18:24:08-06:00] ERROR: Exception handlers complete
[2015-06-07T18:24:08-06:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
Chef Client failed. 33 resources updated in 28.384969 seconds
[2015-06-07T18:24:08-06:00] ERROR: homebrew_package[brew-cask] (sprout-homebrew::cask line 13) had an error: TypeError: no implicit conversion of Fixnum into String
[2015-06-07T18:24:08-06:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

[1]: https://github.com/pivotal-sprout/sprout-homebrew
[2]: http://jtimberman.housepub.org/blog/2014/12/29/chef-12-homebrew-user-mixin/
